### PR TITLE
Refactored model utils.

### PR DESCRIPTION
**Pull request by oskarwrobel**
Fixes https://github.com/ckeditor/ckeditor5-engine/issues/306

Related to:
- https://github.com/ckeditor/ckeditor5-enter/pull/22
- https://github.com/ckeditor/ckeditor5-typing/pull/35
- https://github.com/ckeditor/ckeditor5-undo/pull/33
- https://github.com/ckeditor/ckeditor5-basic-styles/pull/18
- https://github.com/ckeditor/ckeditor5-headings/pull/20

Temporary branches configuration available here: https://github.com/ckeditor/ckeditor5/tree/t/ckeditor5-engine/306

### CODE REVIEW

File: src/dataprocessor/xmldataprocessor.js [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+		/**
+		 * DOM converter used to convert DOM elements to view elements.
+		 *
+		 * @private
+		 * @member {engine.view.DomConverter} engine.dataProcessor.XmlDataProcessor#_domConverter.
+		 */
+		this._domConverter = new DomConverter( { blockFiller: NBSP_FILLER } );
+
+		/**
+		 * BasicHtmlWriter instance used to convert DOM elements to HTML string.
```
Comment (pjasiun) [inline src/dataprocessor/xmldataprocessor.js]: A comment is needed why XML DP use HTML writer.

File: tests/_utils-tests/model.js [hunk 2]
```diff
@@ -195,14 +200,14 @@ describe( 'model test utils', () => {
 		it( 'writes element attributes', () => {
 			root.appendChildren(
 				new Element( 'a', { foo: true, bar: 1, car: false }, [
-					new Element( 'b', { fooBar: 'x y', barFoo: { x: 1, y: 2 } } )
+					new Element( 'b', { fooBar: 'x y', barFoo: JSON.stringify( { x: 1, y: 2 } ) } )
```
Comment (pjasiun) [inline tests/_utils-tests/model.js]: That's strange. It should be possible to set object as an attribute. It should be `parse` who stringify attributes.

File: tests/_utils-tests/model.js [hunk 3]
```diff
@@ -256,32 +261,33 @@ describe( 'model test utils', () => {
...
 			} );
 
 			it( 'writes selection collapsed at the text right boundary', () => {
 				selection.collapse( elB, 'before' );
 
 				expect( stringify( root, selection ) ).to.equal(
-					'<a></a>foo<$text bold=true>bar</$text><selection bold=true /><b></b>'
+					'<a></a>foo<$text bold="true">bar</$text>[]<b></b>'
 				);
+				expect( selection.getAttribute( 'bold' ) ).to.true;
```
Comment (pjasiun) [inline tests/_utils-tests/model.js]: You do not check if there is no other selection attributes. Add check to the number of attributes.

File: tests/_utils-tests/model.js [hunk 4]
```diff
@@ -291,25 +297,17 @@ describe( 'model test utils', () => {
...
 			} );
 
 			it( 'writes selection collapsed selection in a text with attributes', () => {
 				selection.collapse( root, 5 );
 
 				expect( stringify( root, selection ) ).to.equal(
-					'<a></a>foo<$text bold=true>b<selection bold=true />ar</$text><b></b>'
+					'<a></a>foo<$text bold="true">b[]ar</$text><b></b>'
 				);
+				expect( selection.getAttribute( 'bold' ) ).to.true;
```
Comment (pjasiun) [inline tests/_utils-tests/model.js]: As above.

File: tests/_utils-tests/model.js [hunk 5]
```diff
@@ -460,61 +453,61 @@ describe( 'model test utils', () => {
...
-			} ).to.throw( Error, 'Parse error - unexpected closing tag.' );
+			} ).to.throw( Error );
 		} );
 
 		it( 'throws when missing closing tag for text', () => {
 			expect( () => {
 				parse( '<$text>' );
-			} ).to.throw( Error, 'Parse error - missing closing tags: $text.' );
+			} ).to.throw( Error );
 		} );
```
Comment (pjasiun) [inline tests/_utils-tests/model.js]: Do we need all these tests now? Aren't they check native behavior now? Maybe one will be enough then.

File: tests/_utils-tests/model.js [hunk 6]
```diff
@@ -523,58 +516,77 @@ describe( 'model test utils', () => {
...
 			} );
 
 			test( 'sets selection within a text', {
-				data: 'x<selection bold=true>y</selection>z'
+				data: 'x[y]z'
 			} );
 
-			test( 'sets selection within a text with different attributes', {
-				data: '<$text bold=true>fo<selection bold=true>o</$text>ba</selection>r'
+			it( 'sets a backward selection containing an element', () => {
```
Comment (pjasiun) [inline tests/_utils-tests/model.js]: "backward", you said... interesting...

File: tests/_utils/view.js [hunk 7]
```diff
@@ -255,13 +266,17 @@ export function stringify( node, selectionOrPositionOrRange = null, options = {}
  * When set to `null` root element will be created automatically. If set to
  * {@link engine.view.Element Element} or {@link engine.view.DocumentFragment DocumentFragment} - this node
  * will be used as root for all parsed nodes.
+ * @param {Boolean} [options.sameSelectionCharacters=false] When set to `true` then selection inside text should be marked using `{` and `}`
+ * and selection outside text using `[` and `]`. When set to `false` then both should be marked with `[` and `]` only.
```
Comment (pjasiun) [inline tests/_utils/view.js]: Isn't it opposite?

File: src/dataprocessor/xmldataprocessor.js [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+ * For licensing, see LICENSE.md.
+ */
+
+import BasicHtmlWriter from './basichtmlwriter.js';
+import DomConverter from '../view/domconverter.js';
+import { NBSP_FILLER } from '../view/filler.js';
+
+/**
+ * XmlDataProcessor class.
+ * This data processor implementation uses XML as input/output data.
```
Comment (pjasiun) [inline src/dataprocessor/xmldataprocessor.js]: Explanation what is the difference between  XML and HTML data processor would be nice, maybe with some examples. I believe this is a question everybody will ask, checking this class: why we need both HTML and XML DP.

File: src/dataprocessor/xmldataprocessor.js [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+	/**
+	 * Converts XML String to its DOM representation. Returns DocumentFragment, containing nodes parsed from
+	 * provided data.
+	 *
+	 * @private
+	 * @param {String} data
+	 * @returns {DocumentFragment}
+	 */
+	_toDom( data ) {
+		data = `<xml xmlns:attribute="foo" xmlns:container="foo">${ data }</xml>`;
```
Comment (pjasiun) [inline src/dataprocessor/xmldataprocessor.js]: Do we need this `foo` here? Can't it be empty? Also `attribute:x` and `container:x` is a specific testing utils notation. Is should not be here, but in these utils. Additionally we need a comment why we add this tag.

File: src/dataprocessor/xmldataprocessor.js [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+	 * Converts provided {@link engine.view.DocumentFragment DocumentFragment} to data format - in this case XML string.
+	 *
+	 * @param {engine.view.DocumentFragment} viewFragment
+	 * @returns {String} XML string.
+	 */
+	toData( viewFragment ) {
+		// Convert view DocumentFragment to DOM DocumentFragment.
+		const domFragment = this._domConverter.viewToDom( viewFragment, document );
+
+		// Convert DOM DocumentFragment to HTML output.
```
Comment (pjasiun) [inline src/dataprocessor/xmldataprocessor.js]: `to HTML` or `to XML`? Why `getHtml` to get "XML string"?

File: tests/_utils/model.js [hunk 11]
```diff
@@ -3,20 +3,41 @@
  * For licensing, see LICENSE.md.
  */
 
-import TreeWalker from '/ckeditor5/engine/model/treewalker.js';
+import Mapper from '/ckeditor5/engine/conversion/mapper.js';
+import Document from '/ckeditor5/engine/model/document.js';
```
Comment (pjasiun) [inline tests/_utils/model.js]: You could use `ModelDocument`, `ModelRange` and `ModelPosition` not to confuse with native one, especially if you have `ModelElement` below.

File: tests/_utils/model.js [hunk 12]
```diff
@@ -3,20 +3,41 @@
...
+import viewWriter from '/ckeditor5/engine/view/writer.js';
+
+import count from '/ckeditor5/utils/count.js';
+import { parse as viewParse, stringify as viewStringify } from '/tests/engine/_utils/view.js';
+import { convertRangeSelection, convertCollapsedSelection } from '/ckeditor5/engine/conversion/model-selection-to-view-converters.js';
+
+// Test utils uses `<$text foo="bar">Lorem ipsum</$text>` notation to create text with attributes, but `$text` is not
+// valid XML element name, so needs to be parsed before conversion to view.
+const VIEW_TEXT_WITH_ATTRIBUTES_ELEMENT = 'view-text-with-attributes';
+const DATA_STRING_TEXT_WITH_ATTRIBUTES_ELEMENT = '$text';
```
Comment (pjasiun) [inline tests/_utils/model.js]: I don't find this `const` useful, but it might be me ;)

File: tests/_utils/model.js [hunk 13]
```diff
@@ -138,367 +179,223 @@ export function stringify( node, selectionOrPositionOrRange = null ) {
...
-	for ( let patternName in patterns ) {
-		match = data.match( patterns[ patternName ] );
+function insertElement() {
+	return ( evt, data, consumable, conversionApi ) => {
+		consumable.consume( data.item, 'insert' );
 
-		if ( match ) {
-			data = data.slice( match[ 0 ].length );
+		const viewPosition = conversionApi.mapper.toViewPosition( data.range.start );
+		const viewElement = new ViewElement( data.item.name, data.item.getAttributes() );
```
Comment (pjasiun) [inline tests/_utils/model.js]: Use can use generic `insertElement` and add view element constructor as a parameter. Something like this:
```
insertElement( data => new ViewElement( data.item.name, data.item.getAttributes() ) );
```

File: tests/_utils/model.js [hunk 14]
```diff
@@ -138,367 +179,223 @@ export function stringify( node, selectionOrPositionOrRange = null ) {
...
-
-	throw new Error( 'Parse error - unexpected token: ' + data + '.' );
+		conversionApi.mapper.bindElements( data.item, viewElement );
+		viewWriter.insert( viewPosition, viewElement );
+	};
 }
 
-function parseAttributes( attrsString ) {
-	attrsString = attrsString.trim();
+function insertText() {
```
Comment (pjasiun) [inline tests/_utils/model.js]: You should be able to use default `insertText`

File: tests/_utils/model.js [hunk 15]
```diff
@@ -138,367 +179,223 @@ export function stringify( node, selectionOrPositionOrRange = null ) {
...
-		ret += writeItem( value, selection, { selection: withSelection } );
+	modelToView.on( 'insert:$text', insertText() );
+	modelToView.on( 'insert', insertElement() );
+	modelToView.on( 'selection', convertRangeSelection() );
+	modelToView.on( 'selection', convertCollapsedSelection() );
 
-		lastPosition = value.nextPosition;
-	}
+	// Convert model to view.
+	modelToView.convertInsert( range );
```
Comment (pjasiun) [inline tests/_utils/model.js]: https://github.com/ckeditor/ckeditor5-engine/pull/560
```
convertInsert -> convertInsertion
```

File: src/dataprocessor/xmldataprocessor.js (hunk 9 above)
Comment (oskarwrobel) [inline src/dataprocessor/xmldataprocessor.js]: - I've just checked and there must to be some value `xmlns:attribute: Empty XML namespace is not allowed`.
- I'll move this namespaces to config.

File: tests/_utils-tests/model.js (hunk 2 above)
Comment (oskarwrobel) [inline tests/_utils-tests/model.js]: I'll fix it working on https://github.com/ckeditor/ckeditor5-engine/issues/556.

**Pull request merged**